### DISCUSSION TITLE
Is there a way to make a fake serial port that knows when it's disconnected?

### DISCUSSION THREAD

**Discussion by jordemort** [dosemu2/dosemu2#2101] [Q&A]
Hey, I'm another one of those people that wants to run old BBS door games. I've been mostly successful with DOSEMU2 so far, but the one hitch I can't seem to get past is cleanly disconnecting from clients.

Here's what I want to happen: I want to start DOSEMU2 with COM1 connected to a PTY or some sort of socket. When the other end of the connection is closed, I want DOSEMU2 to alter the state of the modem control lines such that whatever is running in DOS acts as if the modem carrier was lost. My knowledge of DOS and serial ports is very sparse, but I believe what this means is that I want DOSEMU2 to switch off the DCD line when the other side of the PTY is closed. I want this so that in case one of my users forcefully kills their SSH session, the DOS program still has a chance to execute its "caller hung up" logic and exit cleanly.

I've tried various ways, with `socat` and with connecting COM1 directly to a PTY, but I can't seem to get the behavior I want. Is it possible?

**Comment by jordemort** [comment]
I think this is the bit of code that I will need to change for it to work the way I want:

https://github.com/dosemu2/dosemu2/blob/5bd19d48d3bc54eb2ffdc5aab07b0973ba5b85a9/src/base/serial/ser_irq.c#L202-L203

Instead of always returning the constant set of "I'm connected!" MSRs, I need to somehow detect that the other side is closed and drop `UART_MSR_DCD` from that set if so.

**Reply by stsp** [reply]
Yes, this is probably possible.
In tty_uart_fill() the read() returning
0 means the other side is closed.
Just add a "closed" member in a com
port struct, and set it there. And in
the code you pointed above, you check
it. Sounds simple.

**Reply by jordemort** [reply]
Hm, I can't quite get it to work. Here's my diff: https://gist.github.com/jordemort/2c5b090b66189304172113e404f241e5

At first I tried just modifying tty_uart_fill and modstat_engine but it didn't seem like tty_uart_fill was getting called unless there was data to actually read on the file descriptor, so I tried adding exceptfds to io_select, but that didn't work either. I don't really understand the control flow here, maybe you can spot something I'm missing?

Log captured with `-D+s` here: https://gist.github.com/jordemort/49133d390095e583c3daf56ea5e3112d

What I did was start a shell inside of qodem and found the pty it was using (/dev/pts/2) - then I started dosemu2 in a separate session with `$_com1 = "/dev/pts/2 pseudo" and launched node 1 of LORD. This works and I am able to interact with LORD in qodem. After that I killed the qodem process, thus closing and destroying the pty that dosemu2 was talking to; it doesn't seem to notice, though, even with my patch; it just sits there until the door times out.

**Reply by stsp** [reply]
Your setup is wrong.
You need to use "exec", not "pseudo"
keyword.

**Reply by jordemort** [reply]
I'm not sure I understand. I'm not trying to get DOSEMU2 to execute something and attach it to the serial port; I'm trying to get it to attach an already existing PTY to the serial port, and notice when the other end of it is closed. I don't think `exec` is what I want for that?

**Reply by stsp** [reply]
Exec opens PTY and the spawned program
opens PTS. This is how the communication
is established.
Legacy scheme where you could open a
pre-configured /dev/ptyXX is not working:
https://cateee.net/lkddb/web-lkddb/LEGACY_PTYS.html
Please stick to "exec".

**Reply by stsp** [reply]
So what you do, is "attaching" dosemu2
to pts. This can be done by specifying the
pts name as you did. But the only way to
open pty (rather than pts) is to use "exec".

**Reply by jordemort** [reply]
Actually someone on Mastodon noticed an error in my patch. I've got it working now. I'll do a PR later today.

**Reply by jordemort** [reply]
Yeah, sorry, definitely not an expert here, got pty mixed up with pts. Qodem is creating the pty for me, and I am connecting DOSEMU2's COM1 to the corresponding pts.

**Reply by stsp** [reply]
When you close pty, then pts users
IIRC get a SIGHUP.  dosemu2 terminates
by a SIGHUP. So I don't believe this is
a viable scheme, but maybe you used
`nohup` or some other trick... Anyway,
"pseudo" keyword is not needed to
access pts, setting its name is enough.

**Reply by stsp** [reply]
Also I think in this case there is no
way to "re-connect". You will need
to create another pts and restart
dosemu2 to reconnect.
So I believe the only viable scheme
is for dosemu to use pty, not pts.

**Comment by jordemort** [comment]
Working patch is here: https://github.com/dosemu2/dosemu2/pull/2102

Closing this to continue discussion over there

**Comment by stsp** [comment]
I now added the "pts" keyword to
allow creating pts and using its pty.
Not sure if it is useful for your task,
but at least now "exec" is not the
only supported way to work with pty.
